tools/decompress-assemblies/main.cs

```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Buffers;
+using System.IO;
+
+using K4os.Compression.LZ4;
+using Xamarin.Tools.Zip;
+
+namespace Xamarin.Android.Tools.DecompressAssemblies
+{
+	class App
+	{
+		const uint CompressedDataMagic = 0x5A4C4158; // 'XALZ', little-endian
+
+		static readonly ArrayPool<byte> bytePool = ArrayPool<byte>.Shared;
+
+		static int Usage ()
+		{
+			Console.WriteLine ("Usage: decompress-assemblies {file.apk|file.dll} [{file.apk|file.dll} ..]");
+			Console.WriteLine ();
+			Console.WriteLine ("Files are uncompressed to the current directory with the `uncompressed-` prefixed to their name.");
+			return 1;
+		}
+
+		static bool UncompressDLL (Stream stream, string filePath)
+		{
+			string outputFile = $"uncompressed-{Path.GetFileName (filePath)}";
+			bool retVal = true;
+
+			Console.WriteLine ($"Processing {filePath}");
+			//
+			// LZ4 compressed assembly header format:
+			//   uint magic;                 // 0x5A4C4158; 'XALZ', little-endian
+			//   uint descriptor_index;      // Index into an internal assembly descriptor table
+			//   uint uncompressed_length;   // Size of assembly, uncompressed
+			//
+			using (var reader = new BinaryReader (stream)) {
+				uint magic = reader.ReadUInt32 ();
+				if (magic == CompressedDataMagic) {
+					reader.ReadUInt32 (); // descriptor index, ignore
+					uint decompressedLength = reader.ReadUInt32 ();
+
+					int inputLength = (int)(stream.Length - 12);
+					byte[] sourceBytes = bytePool.Rent (inputLength);
+					reader.Read (sourceBytes, 0, inputLength);
+
+					byte[] assemblyBytes = bytePool.Rent ((int)decompressedLength);
+					int decoded = LZ4Codec.Decode (sourceBytes, 0, inputLength, assemblyBytes, 0, (int)decompressedLength);
+					if (decoded != (int)decompressedLength) {
+						Console.Error.WriteLine ($"  Failed to decompress LZ4 data of {filePath} (decoded: {decoded})");
+						retVal = false;
+					} else {
+						using (var fs = File.Open (outputFile, FileMode.Create, FileAccess.Write)) {
+							fs.Write (assemblyBytes, 0, assemblyBytes.Length);
+							fs.Flush ();
+						}
+						Console.WriteLine ($"  uncompressed to: {outputFile}");
+					}
+
+					bytePool.Return (sourceBytes);
+					bytePool.Return (assemblyBytes);
+				} else {
+					Console.WriteLine ($"  assembly is not compressed");
+				}
+			}
+
+			return retVal;
+		}
+
+		static bool UncompressDLL (string filePath)
+		{
+			using (var fs = File.Open (filePath, FileMode.Open, FileAccess.Read)) {
+				return UncompressDLL (fs, filePath);
+			}
+		}
+
+		static bool UncompressFromAPK (string filePath)
+		{
+			using (ZipArchive apk = ZipArchive.Open (filePath, FileMode.Open)) {
+				foreach (ZipEntry entry in apk) {
+					if (!entry.FullName.StartsWith ("assemblies/", StringComparison.Ordinal)) {
```

Comment:
I think if we changed this one line to be `Contains` rather than `StartsWith` this tool could support `aab` files as well.
(based on my knowledge of aab files). 
@jonathanpeppers my understanding is the assemblies in `aab` will be in `root/assemblies` correct?

Reply:
@dellis1972 good call! Support for AAB added :)